fix: insert batch rewrite.

Chrriis opened the pull request:
I added a new test which does not pass because batch rewrite assumes the values to add have a number of parameters equal to the number of dynamic parameters. This omits the existence of fixed parameters.
So a statement like:
INSERT INTO testbatch VALUES (?,1)
becomes:
INSERT INTO testbatch VALUES ($1,1),($2),($3)
I guess it should replicate the whole block, which should become:
INSERT INTO testbatch VALUES ($1,1),($2,1),($3,1)

vlsi commented:
Looks good.
Would you implement the fix as well?

Chrriis commented:
I might give it a try but I am not sure yet what the best approach is for a fix.

I have the feeling that in BatchedQueryDecorator:246, we should find in "s" whatever is after the "VALUES" token and replicate that as many times as needed. If I am not mistaken, we cannot have anything fancy after the VALUES token, because we only get a BatchedQueryDecorator if it is simple enough to be rewritten.

vlsi commented:
> we cannot have anything fancy after the VALUES token

Technically speaking, there's `?, ? -> $1, $2` kind of translation somewhere, so it might be a bit tricky to replace `?s` properly.

Chrriis commented:
I added a fix. From what I understood, only a single set of parentheses is allowed for the batch to be rewritten. So I can safely consider the last set of parentheses to be the block of values. I then rewrite that block, renumbering the parameters using the expected index.

vlsi commented on pgjdbc/src/main/java/org/postgresql/core/v3/BatchedQueryDecorator.java:
```diff
@@ -240,17 +240,30 @@ String getNativeSql() {
+        for (int j = startIndex; j < endIndex; j++) {
+          char ch = s.charAt(j);
+          switch (ch) {
+            case '$':
```
Will it blend for `values (?, 'dollar here: $', ?)` kind of statement?

vlsi commented:
Ah, it seems to require more involved parsing so `values (?, 'literals with ?, (, $, and )', ?)` works.

Chrriis commented:
In fact you are right. These do not pass:
VALUES (?, 'literals with ?, (, $1234, and )', ?)
VALUES ($1,'1, (, $1234, a''n?d )') /*xxxx)*/

I can improve but this is a slippery slope into writing a complete parser.

vlsi commented:
>I can improve but this is a slippery slope into writing a complete parser.

Do you realize pgjdbc has to follow that route for each and every `PreparedStatement`?
It has to somehow parse the SQL and replace `?` bind placeholders with `$n` style ones (with account for  comments and literals).
Frankly speaking, that was a bit surprising for me when I realized that for a first time.

Having said that, I see the following options:
1) Add test (and more extreme variations) with `@Ignore` annotation, so somebody could pick it up later
2) Teach batch rewriter to rewrite only those statements that are simple `values (?,?,?)` (that is just questionmarks commas and spaces after `values`)
3) Reuse `postgresql.core.Parser` and rewrite with account for comments/literals.

polobo commented:
On Tue, Jun 7, 2016 at 12:57 PM, Vladimir Sitnikov <notifications@github.com
> wrote:

> I can improve but this is a slippery slope into writing a complete parser.
>
> Do you realize pgjdbc has to follow that route for each and every
> PreparedStatement?
> It has to somehow parse the SQL and replace ? bind placeholders with $n
> style ones (with account for comments and literals).
> Frankly speaking, that was a bit surprising for me when I realized that
> for a first time.
>
> Having said that, I see the following options:
> 1) Add test (and more extreme variations) with @Ignore annotation, so
> somebody could pick it up later
> 2) Teach batch rewriter to rewrite only those statements that are simple values
> (?,?,?) (that is just questionmarks commas and spaces after values)
> 3) Reuse postgresql.core.Parser and rewrite with account for
> comments/literals.
>
FTR I'd be concerned about unintended consequences while doing #2 but from
a usability perspective I would not complain if I had to operate within
that restriction.  Pulling constants out and placing them in the client
Java code is trivial and you already have one parameter you are dealing
with so the overhead is sunk.

Also, these should not be considered mutually exclusive.​

David J.
​

Chrriis commented:
> Do you realize pgjdbc has to follow that route for each and every PreparedStatement?

Yes I do :)
What a pity that the parsing it did has to be done again in case of the batch rewriting.

About 1), of course, this is something I should do.

About 2), where I work we use comments in queries, because we do query rewrite to target alternate views than the indicated tables under certain conditions. So ignoring the case of comments would greatly reduce the usefulness of that feature for us.

I am trying 3) but I see that certain parsing like parseSingleQuotes require parameters which I am not sure where to find if I were to replicate such parsing in the context of the batch rewriter.
I am also tempted to keep track of certain key positions from the initial parsing, but then I am not sure where this data should go to be picked up again at rewriting time.

vlsi commented:
>I am also tempted to keep track of certain key positions from the initial parsing

Frankly speaking, "batch rewriter" is in my "to review" list. I'm not familiar with that code.

>I am also tempted to keep track of certain key positions from the initial parsing, 

So am I, however I'm hot sure if that's feasible. On the other hand, it looks like `int valuesBraceOpenPosition, valuesBraceClosePosition; int[] questionMarkPositions` should be enough to cerate proper multivalues.

Chrriis commented:
> On the other hand, it looks like int valuesBraceOpenPosition, valuesBraceClosePosition; int[] questionMarkPositions should be enough to cerate proper multivalues.

This is exactly what I am trying to do.
In BatchedQueryDecorator, I have access to "getNativeQuery().bindPositions", which seem to be the equivalent of "questionMarkPositions".
I need "valuesBraceOpenPosition, valuesBraceClosePosition", which I might be able to extract/compute in QueryExecutorImpl to pass to the BatchedQueryDecorator.

Chrriis commented:
I just pushed some code to fix various issues. I had to go a bit further than I had expected, so please let me know if I did things in unexpected ways.
So now, during parsing, we don't set a flag to indicate batch rewrite anymore, but we set the start and end positions of the values parameters block (start/end parentheses). Because parsing already stored the positions of the original parameters to bind and we know the start and end of the block, we know exactly which parts of the query to replicate (the chunks around the parameters within the values block).
I also improved parsing because the keyword detection logic was creating too many false positive (for example "AVALUES" entered the "VALUES" logic).

Chrriis commented:
Note: I don't know what to do with the test "DeepBatchedInsertStatementTest.testNativeSqlSizeCalculation" because size calculation needs to know the block information now.

Chrriis commented:
On my real world setup, the new code fails, but not because of parsing. The general idea is that every batch adds a new block of parameters, but there is no place where it says it should stop and start a new batch. The first of my real world cases has:
- 22 dynamic parameters, 1 fixed parameter.
- 16'450 batches.
- Which means 361'900 parameters after rewrite.
- The resulting statement is 3'212'254 characters long.

And I get the error: SQL State 08006
"java.io.IOException: Tried to send an out-of-range integer as a 2-byte value: 361900"

I guess Postgres has an upper limit of number of parameters around 65536 (if unsigned). In any case, I think Postgres performance is best with about 100 blocks of values (not 16450!) so we should aggregate with max of 100 (or any magical number if it should depend on number of parameters). In my original external logic (that is before I tried to fix the driver feature), I used a formula like: "Math.min((2000 - 1) / parameterCount, 100)".

I am not sure how to start a new batch every 100 parameters, and I am not even sure it is possible. If it is not possible to divide a query into multiple queries, then the whole feature can be thrown away. Any ideas?

vlsi commented:
> I am not sure how to start a new batch every 100 parameters, and I am not even sure it is possible

Well, we should split on `values` boundaries. That should work.

> I think Postgres performance is best with about 100 blocks of values

According to https://github.com/pgjdbc/pgjdbc/pull/491#issuecomment-196881964, the gain from 16 -> 128 values is not that high, so having something like "`Math.min(100, 65535/num_binds)` values max" makes sense.

vlsi commented:
>I guess Postgres has an upper limit of number of parameters around 65536 

>https://www.postgresql.org/docs/9.5/static/protocol-message-formats.html
>Int16
>The number of parameter format codes that follow (denoted C below). This can be zero to indicate that there are no parameters or that the parameters all use the default format (text); or one, in which case the specified format code is applied to all parameters; or it can equal the actual number of parameters.

The number of parameters is indeed just int16. Not sure if that's signed or not from backend's point of view.

Chrriis commented:
> I am not sure how to start a new batch every 100 parameters

I actually meant every 100 blocks of parameters of course.
What I meant was that I am not sure if a new batch query can be created every 100 blocks and added to the list of queries to be processed. I will have to investigate, though I am not that familiar with the code yet.

> The number of parameters is indeed just int16. Not sure if that's signed or not from backend's point of view.

I would like to limit to some reasonable number which does not result in huge statements. This number should be way below those technical limit to not have to ask whether it is signed or not :)

whitingjr commented on pgjdbc/src/main/java/org/postgresql/core/Parser.java:
```diff
@@ -416,122 +412,6 @@ public static int parseBlockComment(final char[] query, int offset) {
-   * @param offset position of query to start checking
-   * @return boolean indicates presence of word
-   */
-  public static boolean parseDeleteKeyword(final char[] query, int offset) {
```
Please don't substitute the code to coerce the char. Dave indicated this technique is necessary.
https://github.com/pgjdbc/pgjdbc/pull/491#issuecomment-209350929

whitingjr commented:
@Chrriis thank you.

whitingjr commented:
@Chrriis I've been thinking about how to build in the splitting up of an individual statement. Avoiding too many values excepitons. At the moment the driver has the concept of a multi statement. org.postgresql.jdbc.PgStatement.batchStatements at the moment will only contain a single statement. This allows the ability to have more than one.
 It should be possible to detect too many entries and build another statement. The QueryExecutorImpl will need updating to issue the extra statements to the back end.

Chrriis commented:
My last commit is a version that works in our production environment. My real world test case went down from 54 seconds to 28 seconds! This is better than my external implementation with which I got down to 32 seconds.

That commit handles batches of plain statements, which was not handled by the original implementation and broke with my test case (yes, it has those too). That test case is a whole sequence of select/update/insert statements (260 statements in total). It contains 18 batched insert statements, some with 23, 248, 1412, 16450, 78798 batches. Queries also have comments, static parameters, ...
If I am not mistaken, the max number of batches is actually a signed int16. But I decided to find the appropriate number of batches using my formula ("Math.min((2000 - 1) / paramCount, 100)") to avoid sql insanely long.

The final test for me would be to integrate a snapshot build containing those changes and see how our nightly integration tests work, but I have to say I am much more confident they would pass now. Please let me know if you intend to integrate these changes and make such snapshot that I could try.

Chrriis commented:
I improved the test case, which I originally posted on the mailing list, to serve as a benchmarking tool. Here it is:
[MultiInsertTest.zip](https://github.com/pgjdbc/pgjdbc/files/307556/MultiInsertTest.zip)

The results on my machine are (1 million rows, 11 columns, 10 parameters) :
Execution WITHOUT rewrite: 1min 14s 909ms
Execution WITH rewrite: 29s 349ms

I would say, 2.5x boost is impressive! Feel free to play with it and let me know what you find/think.

Note that I had to upload as a zip, but it only is a single standalone test that only requires Postgres driver JAR in the classpath as well as tweaking a few constants at the top of the file.

vlsi commented:
@Chrriis , JFYI: we use JMH for benchmarks: https://github.com/pgjdbc/pgjdbc/blob/master/ubenchmark/src/main/java/org/postgresql/benchmark/statement/InsertBatch.java

Chrriis commented:
@vlsi Thanks, I did not know. That being said, this does not seem to fit with what I was doing: I was showing the gains, not just measuring the current state of things. Correct me if I am wrong, but the test you pointed me to does not measure gains between 2 approaches.

vlsi commented:
>Correct me if I am wrong, but the test you pointed me to does not measure gains between 2 approaches

To measure "before vs after" one just needs to run the same benchmark against different pgjdbc versions.

>I was showing the gains

The gains might depend on "the number of columns", "number of rows", etc, etc. It turns out JMH allows walking the parameter space, so we can check if there is a sweet point of "100 binds" or "100 rows" or whatever.

Chrriis commented:
@vlsi I think we agree but we were not on the same train of thoughts :)
I was merely providing the order of magnitude of gains, what to expect by having this feature active.
As a follow up on my findings, for 10 million rows I have more than 3 times gain (6 minutes instead of 20).

vlsi commented:
```sql
  public void testBindsInNestedParens()
      throws SQLException {
    PreparedStatement pstmt = null;
    try {
      pstmt = con.prepareStatement("INSERT INTO testbatch VALUES ((?),((?)),?);");
      pstmt.setInt(1, 1);
      pstmt.setString(2, "a");
      pstmt.setInt(3, 2);
      pstmt.addBatch();
      pstmt.setInt(1, 3);
      pstmt.setString(2, "b");
      pstmt.setInt(3, 4);
      pstmt.addBatch();
      pstmt.setInt(1, 5);
      pstmt.setString(2, "c");
      pstmt.setInt(3, 6);
      pstmt.addBatch();
      assertTrue(
          "Three batched statements should be rewritten to single multi-values one, thus expecting SUCCESS_NO_INFO",
          Arrays.equals(new int[] { Statement.SUCCESS_NO_INFO,
              Statement.SUCCESS_NO_INFO, Statement.SUCCESS_NO_INFO },
              pstmt.executeBatch()));
    } finally {
      TestUtil.closeQuietly(pstmt);
      con.rollback();
    }
  }
```

triggers
```
java.lang.StringIndexOutOfBoundsException: String index out of range: -5

	at java.lang.String.substring(String.java:1967)
	at org.postgresql.core.v3.BatchedQueryDecorator.getNativeSql(BatchedQueryDecorator.java:273)
	at org.postgresql.core.v3.QueryExecutorImpl.sendParse(QueryExecutorImpl.java:1306)
	at org.postgresql.core.v3.QueryExecutorImpl.sendOneQuery(QueryExecutorImpl.java:1648)
	at org.postgresql.core.v3.QueryExecutorImpl.sendQuery(QueryExecutorImpl.java:1232)
	at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:367)
	at org.postgresql.jdbc.PgStatement.executeBatch(PgStatement.java:1016)
	at org.postgresql.test.jdbc2.BatchedInsertReWriteEnabledTest.testBindsInNestedParens(BatchedInsertReWriteEnabledTest.java:189)
```

vlsi commented:
This one is tricky as well:
```sql
INSERT INTO testbatch (pk) VALUES (?), VALUES (?)
```

We either can skip rewritting multi-values inserts, or use `valuesStartPosition` instead of `valuesBraceOpenPosition` (that is replicate all the thing from the first VALUES till the closing paren)

vlsi commented:
This one fails: it tries to batch 32000 binds into two-valued insert and fails due to "sendInt16" assumes <=32768
```sql
  public void test32000Binds() throws Exception {
    PreparedStatement pstmt = null;
    try {
      StringBuilder sb = new StringBuilder();
      sb.append("INSERT INTO testbatch(pk) VALUES (coalesce(?");
      for (int i = 0; i < 32000 - 1 /* note one ? above */; i++) {
        sb.append(",?");
      }
      sb.append("))");
      pstmt = con.prepareStatement(sb.toString());
      for (int k = 0; k < 2; k++) {
        for (int i = 1; i <= 32000; i++) {
          pstmt.setInt(i, i);
        }
        pstmt.addBatch();
      }
      assertEquals(
          "Statement with 60000 binds should not be batched => two executions with exactly one row inserted each",
          Arrays.toString(new int[] { 1, 1 }),
          Arrays.toString(pstmt.executeBatch()));
    } catch(BatchUpdateException be) {
      SQLException e = be;
      while (true) {
        e.printStackTrace();
        SQLException next = e.getNextException();
        if (next == null) {
          break;
        }
        e = next;
      }
      throw e;
    } finally{
      TestUtil.closeQuietly(pstmt);
    }
  }
```

```
Caused by: java.io.IOException: Tried to send an out-of-range integer as a 2-byte value: 64000
	at org.postgresql.core.PGStream.SendInteger2(PGStream.java:214)
	at org.postgresql.core.v3.QueryExecutorImpl.sendParse(QueryExecutorImpl.java:1345)
	at org.postgresql.core.v3.QueryExecutorImpl.sendOneQuery(QueryExecutorImpl.java:1648)
	at org.postgresql.core.v3.QueryExecutorImpl.sendQuery(QueryExecutorImpl.java:1232)
	at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:367)
	... 25 more
```

vlsi commented:
@Chrriis , I've made some changes while doing the review (see https://github.com/pgjdbc/pgjdbc/tree/multivalues_rewriter).

It looks somewhat OK, however `BatchFailureTest` fails when run in `rewrite insert=true` mode (6 failures of 186 cases total). It looks like `BatchResultHandler` should be updated to account for `multi-valued` statements.

Can you please have a look?

The error message is like (`-3 == Statement.EXECUTE_FAILED`)
```
java.lang.AssertionError: Batch should fail at row 503, thus at least 503 items should be returned, actual result is 11 items, [-3, -3, -3, -3, -3, -3, -3, -3, -3, -3, -3]
```

In other words, super-statements did not expand to `int[] batchResult`

vlsi commented:
Regarding "batch result expand logic", I think https://github.com/pgjdbc/pgjdbc/commit/c06ad8ca1d1c6ef3ff11c8415dc7583d5fd6c5a8#diff-e517b352b59b7d012c0a772c7e84d3eeR1026 should be moved to `BatchResultHandler`

vlsi commented:
Note to self: do not merge unless performance tests are run & ok

vlsi commented:
I've updated https://github.com/pgjdbc/pgjdbc/tree/multivalues_rewriter branch.
Notable changes are:
1) `values(` case is now rewritten (the case without space after `values`)
2) `autoCommit(true)` is rewritten as well. I've absolutely no idea why `autoCommit(true)` was a showstopper for insert rewrite. @whitingjr any clue?

Here are the measurements (p1nrows is the number of rows inserted, p2multi is the number of values in the prepared statement):
1) p1nrows <= 128 cases (insert all the rows at once) shows that `reWriteBatchedInserts=true` does not affect response times. That is good
2) p2nrows = 1024 shows that "superbatching" does **NOT** improve the response times much.
For instance: typical `insert ... values()` of 1024 rows takes 8.8 +- 0.5ms with regular approach and 6.2 +- 0.4ms with new approach.

Frankly speaking, I would expect (1024,1) and (1024, 16) to have the same performance in `rewrite=true` case (they both should be rewritten into 11 100-values statements).

```
Benchmark                               (p1nrows)  (p2multi)  Mode  Cnt       Score        Error   Units
insertBatch                                    16        128  avgt   10       0,181 ±      0,008   ms/op
insertBatch:·gc.alloc.rate.norm                16        128  avgt   10    5474,984 ±    141,080    B/op
insertBatch                                   128        128  avgt   10       0,430 ±      0,033   ms/op
insertBatch:·gc.alloc.rate.norm               128        128  avgt   10   32859,756 ±    102,335    B/op
insertBatch                                  1024          1  avgt   10       8,815 ±      0,476   ms/op
insertBatch:·gc.alloc.rate.norm              1024          1  avgt   10  584721,886 ±   4206,864    B/op
insertBatch                                  1024          4  avgt   10       4,048 ±      0,195   ms/op
insertBatch:·gc.alloc.rate.norm              1024          4  avgt   10  334393,775 ±    428,785    B/op
insertBatch                                  1024          8  avgt   10       3,072 ±      0,147   ms/op
insertBatch:·gc.alloc.rate.norm              1024          8  avgt   10  294039,555 ±    348,654    B/op
insertBatch                                  1024         16  avgt   10       2,708 ±      0,156   ms/op
insertBatch:·gc.alloc.rate.norm              1024         16  avgt   10  274398,802 ±    269,792    B/op
insertBatch                                  1024        128  avgt   10       2,359 ±      0,114   ms/op
insertBatch:·gc.alloc.rate.norm              1024        128  avgt   10  257610,558 ±    210,756    B/op
insertBatchWithRewrite                         16        128  avgt   10       0,184 ±      0,017   ms/op
insertBatchWithRewrite:·gc.alloc.rate.norm     16        128  avgt   10    5192,701 ±    113,898    B/op
insertBatchWithRewrite                        128        128  avgt   10       0,429 ±      0,021   ms/op
insertBatchWithRewrite:·gc.alloc.rate.norm    128        128  avgt   10   32876,912 ±     96,789    B/op
insertBatchWithRewrite                       1024          1  avgt   10       6,257 ±      0,389   ms/op
insertBatchWithRewrite:·gc.alloc.rate.norm   1024          1  avgt   10  600380,529 ±    453,958    B/op
insertBatchWithRewrite                       1024          4  avgt   10       4,834 ±      0,207   ms/op
insertBatchWithRewrite:·gc.alloc.rate.norm   1024          4  avgt   10  438489,859 ±    813,896    B/op
insertBatchWithRewrite                       1024          8  avgt   10       3,334 ±      0,109   ms/op
insertBatchWithRewrite:·gc.alloc.rate.norm   1024          8  avgt   10  357850,690 ±    269,867    B/op
insertBatchWithRewrite                       1024         16  avgt   10       2,530 ±      0,132   ms/op
insertBatchWithRewrite:·gc.alloc.rate.norm   1024         16  avgt   10  314700,598 ±    332,069    B/op
insertBatchWithRewrite                       1024        128  avgt   10       2,531 ±      0,159   ms/op
insertBatchWithRewrite:·gc.alloc.rate.norm   1024        128  avgt   10  308652,031 ±    334,311    B/op
```

Here are the results for 9.4.1208 (it looks like rewrite=true did not work in 1208 for some reason as the response times and memory allocated are the same as with rewrite=false case):
```
Benchmark                               (p1nrows)  (p2multi)  Mode  Cnt       Score        Error   Units
insertBatch                                    16        128  avgt   10       0,193 ±      0,030   ms/op
insertBatch:·gc.alloc.rate.norm                16        128  avgt   10    5449,291 ±    225,758    B/op
insertBatch                                   128        128  avgt   10       0,491 ±      0,121   ms/op
insertBatch:·gc.alloc.rate.norm               128        128  avgt   10   32851,079 ±    105,972    B/op
insertBatch                                  1024          1  avgt   10       8,874 ±      0,501   ms/op
insertBatch:·gc.alloc.rate.norm              1024          1  avgt   10  576482,744 ±   3019,834    B/op
insertBatch                                  1024          4  avgt   10       4,033 ±      0,307   ms/op
insertBatch:·gc.alloc.rate.norm              1024          4  avgt   10  332328,831 ±    510,827    B/op
insertBatch                                  1024          8  avgt   10       3,064 ±      0,129   ms/op
insertBatch:·gc.alloc.rate.norm              1024          8  avgt   10  293029,055 ±    283,119    B/op
insertBatch                                  1024         16  avgt   10       2,737 ±      0,213   ms/op
insertBatch:·gc.alloc.rate.norm              1024         16  avgt   10  273889,446 ±    259,054    B/op
insertBatch                                  1024        128  avgt   10       2,353 ±      0,165   ms/op
insertBatch:·gc.alloc.rate.norm              1024        128  avgt   10  257513,918 ±    365,662    B/op
insertBatchWithRewrite                         16        128  avgt   10       0,179 ±      0,019   ms/op
insertBatchWithRewrite:·gc.alloc.rate.norm     16        128  avgt   10    5467,017 ±    140,891    B/op
insertBatchWithRewrite                        128        128  avgt   10       0,423 ±      0,031   ms/op
insertBatchWithRewrite:·gc.alloc.rate.norm    128        128  avgt   10   32851,484 ±    103,904    B/op
insertBatchWithRewrite                       1024          1  avgt   10       8,642 ±      0,343   ms/op
insertBatchWithRewrite:·gc.alloc.rate.norm   1024          1  avgt   10  576117,425 ±   1267,016    B/op
insertBatchWithRewrite                       1024          4  avgt   10       4,090 ±      0,332   ms/op
insertBatchWithRewrite:·gc.alloc.rate.norm   1024          4  avgt   10  332571,721 ±   1174,038    B/op
insertBatchWithRewrite                       1024          8  avgt   10       3,171 ±      0,640   ms/op
insertBatchWithRewrite:·gc.alloc.rate.norm   1024          8  avgt   10  293010,584 ±    372,193    B/op
insertBatchWithRewrite                       1024         16  avgt   10       2,684 ±      0,112   ms/op
insertBatchWithRewrite:·gc.alloc.rate.norm   1024         16  avgt   10  273894,537 ±    233,297    B/op
insertBatchWithRewrite                       1024        128  avgt   10       2,350 ±      0,125   ms/op
insertBatchWithRewrite:·gc.alloc.rate.norm   1024        128  avgt   10  257526,558 ±    305,421    B/op
```

vlsi commented:
It turns out in `rewrite=true` pgjdbc is doing lots of `parse` calls. prepared query reuse seems to be broken somehow.

<img width="1113" alt="insert_batch_rewritten_parses" src="https://cloud.githubusercontent.com/assets/213894/15991644/4060dc08-30c1-11e6-8b58-5deae7b4aecf.png">

vlsi commented:
RE "query reuse", there's an issue: different number of rows require different sql texts (with varying number of values clauses).

I see two approaches:
1) split after 100 rows, and create customized statement for the rest. For instance, 190 rows would take 2 queries: 100 rows and 90. This is the current implementation.
Pros: small number of queries
Cons: up to 100 variations of "the same" sql at JDBC and DB sides, thus it worst case might degradation badly.
2) split on power of two. For instance, 190 = 128 + 32 + 16 + 8 + 4 + 2, thus use 6 queries.
Pros: good server-prepared statement reuse. It seems optimal in the worst case.
Cons: more queries than in previous approach, thus it can be slower.

I'm inclined to option 2.

@davecramer , @whitingjr , @Chrriis any ideas on 1 vs 2?

Chrriis commented:
@vlsi I am just back from a very exhausting week-end so it will take me a few days to digest all your posts :)

That being said, I can already answer you last question on how to split batches. And my answer is:
3) Use the batch of n rows converted as values (using my formulas capped to 100), and for the remainder use a regular batch of the original single-values statement. The worst case is if you have 199 rows, you get 1 multi-values statements of 100 values blocks, and 99 single-values statements. In practice, this is not noticeable as the algorithm's power really comes into play when you have several thousands or millions of rows, so the remainder is meaningless.
This is the approach I used for our external implementation and I got similar results as yours except only 2 variations of a batch were involved: the multi-values one and the original one.

Note: the difference of timing I mentioned earlier for my real world case was because I had not optimized one query with multi-insert rewrite. After implementing, I did get similar times.

Chrriis commented on pgjdbc/src/test/java/org/postgresql/test/jdbc2/BatchedInsertReWriteEnabledTest.java:
```diff
@@ -153,16 +255,39 @@ public void testBatchWithReWrittenBatchStatementWithSemiColon()
-              Statement.SUCCESS_NO_INFO, Statement.SUCCESS_NO_INFO },
-              pstmt.executeBatch()));
+      Assert.assertEquals(
+          "Three batched statements should be rewritten to single multi-values one, thus expecting SUCCESS_NO_INFO",
```
@vlsi I don't understand this test case. The original idea was to not perform any rewrite if there is more than 1 opening or closing parenthesis after the VALUES clause.

Chrriis replied:
Actually, this original limitation was because I did not know how code like this would behave:
INSERT INTO X VALUES (?, random(), ?)
... but it seems that a batch of these does not evaluate the function only once. So there is no problem in nesting parentheses.

vlsi replied:
>I don't understand this test case.

Do you mean the assert message should be clarified or you just did not expect the case with parens to be rewritten?

> So there is no problem in nesting parentheses.

I think so.

vlsi commented on pgjdbc/src/test/java/org/postgresql/test/jdbc2/BatchedInsertReWriteEnabledTest.java:
```diff
@@ -82,47 +126,168 @@ public void testBatchWithReWrittenRepeatedInsertStatementOptimizationEnabled()
+  }
+
+  /**
+   * Test to make sure a statement with a semicolon is not broken
```
This is misleading. Test name is fine, yet the comment was copy-pasted from somewhere else.

Chrriis commented:
So you know, I tested those latest changes made by @vlsi against my real world test case and it worked perfectly.

vlsi commented on pgjdbc/src/main/java/org/postgresql/core/v3/BatchedQueryDecorator.java:
```diff
@@ -151,28 +148,6 @@ boolean isPreparedFor(int[] paramTypes) {
-      super.setStatementName(statementName);
-      batchedEncodedName = Utils.encodeUTF8(statementName);
-    }
-  }
```
Awesome

Chrriis replied:
I don't know if it became useless because of my descriptor recycling approach or something else, but it does not seem to be any use as far as I can tell.

Chrriis commented:
@vlsi I am done with that round of changes, so feel free to change whichever way you like and I would pick the changes from your branch. Of course, testing is welcome as well as performance tests to compare with past results.

Chrriis commented:
I did a change which I was not sure to push to the branch of this pull request, so I created another branch: https://github.com/Chrriis/pgjdbc/tree/ReWrittenBatch2
The change is to perform query rewriting at execution time. Before, we were rewriting the parameters at execution time (in `pstmt.transformParameters()`) but we were rewriting the queries in `addBatch()` (splitting when max/100 is reached). Now all the logic is done in `pstmt.transformQueriesAndParameters()` called by `executeBatch()`.
The benefits are that code is in less places, the decorated query behaves like a normal query until execution. Moreover, because we know exactly how many batches we are rewriting, we could decide to improve splitting of remainders (you suggested powers of 2).
Please let me know if you see problems with that approach, if you like it or have suggestions, and if I should merge it to the branch of this pull request.

vlsi commented:
@Chrriis , I like the refactor in ReWrittenBatch2. Can you please incorporate it here?

vlsi commented:
@Chrriis It would be good if you could implement powers of 2.
Note: the idea is that each kind of statement (e.g. the one with 64 repetitions) can be server-prepared. Typically, in order to server-prepare a query, `org.postgresql.jdbc.PgConnection#borrowQuery` kind of API should be used and no "unprepare" kind of calls should be used.


In order to check how batching behaves, you can run some batching unittest with `-Dloglevel=2`. It would create `postgresql-jdbc-tests.debug.txt` log with trace of frontend-backend communication (the idea is to end up with `parse; bind; exec; bind; exec; bind; exec; sync` kind of trace)

Chrriis commented:
@vlsi Thanks a lot for the directions, this was very helpful!

Getting the caching to work was actually tricky. I changed the approach a little: the original decorator is now left untouched and represents the query for one VALUES block. If a different number of blocks is needed when rewriting, then it is obtained from and cached within this original decorator (cf. `BatchedQueryDecorator.deriveForMultiBatch(int)`). So from an LRU cache point of view, only the original decorator is cached, but when rewriting we get the derived forms from that original decorator. I couldn't plug the derived decorators in the LRU cache because caching keys are the statements before transformation (which means `?` instead of `$n` for example).

I also implemented powers of 2, and I bumped the maximum number of blocks to 128 to simplify a little. So all blocks are powers of 2. For example: 451 rows = 128, 128, 128, 64, 2, 1.

It would be great if you could crosscheck that caching works as expected and that performance is good now.

vlsi commented:
The measurements look promising:
```
Benchmark                           (p1nrows)  (p2multi)  Mode  Cnt  Score   Error  Units
InsertBatch.insertBatch                    16        128  avgt    5  0,184 ± 0,041  ms/op
InsertBatch.insertBatch                   128        128  avgt    5  0,471 ± 0,061  ms/op
InsertBatch.insertBatch                  1024          1  avgt    5  9,022 ± 2,171  ms/op
InsertBatch.insertBatch                  1024          4  avgt    5  4,098 ± 0,239  ms/op
InsertBatch.insertBatch                  1024          8  avgt    5  3,313 ± 0,715  ms/op
InsertBatch.insertBatch                  1024         16  avgt    5  2,985 ± 1,282  ms/op
InsertBatch.insertBatch                  1024        128  avgt    5  2,424 ± 0,681  ms/op
InsertBatch.insertBatchWithRewrite         16        128  avgt    5  0,179 ± 0,028  ms/op
InsertBatch.insertBatchWithRewrite        128        128  avgt    5  0,420 ± 0,053  ms/op
InsertBatch.insertBatchWithRewrite       1024          1  avgt    5  2,487 ± 0,149  ms/op
InsertBatch.insertBatchWithRewrite       1024          4  avgt    5  2,503 ± 0,587  ms/op
InsertBatch.insertBatchWithRewrite       1024          8  avgt    5  2,565 ± 0,136  ms/op
InsertBatch.insertBatchWithRewrite       1024         16  avgt    5  2,600 ± 0,086  ms/op
InsertBatch.insertBatchWithRewrite       1024        128  avgt    5  2,596 ± 0,353  ms/op
```

vlsi commented:
We need more functional tests for batch insert: see #583

Chrriis commented:
I was performing all sorts of real-world tests, one of which connects to databases from SQL Server which I read and convert to create an equivalent Postgres database. This is purely based on batched parameterized insert statements involving various tables and column types.
Those that work are much faster (yeah!) but one freezes (ouch!). I cannot figure out if it is because of batch rewrite or if my test in batch rewrite mode activates the necessary conditions of an existing bug. In any case, I need your help.
It took me a long time to isolate the bug to a self contained example. When I investigated, I understood that the Prepared statement types do not match the new types of the batch to execute, so it prepares it with the new types, and eventually freezes when it reaches the 32nd name. I tried to remove all sorts of type fiddling that came from the original implementation (because I am not sure the latest approach needs them) but it did not solve the freeze.

[MultiInsertTest4.zip](https://github.com/pgjdbc/pgjdbc/files/321829/MultiInsertTest4.zip)

All you have to do is edit the top constants, run. You should see "Before... After" sequences, but on my system it stays stuck on "Before...".

Chrriis commented:
I managed to reproduce the freeze with regular prepared statements. cf issue #584

vlsi commented:
I've pushed https://github.com/vlsi/pgjdbc/tree/multivalues_rewriter with my review changes.
The change I like most is the removal of `statementType/statementName` dance from `BatchedQuery` => it should rely on existing machinery for prepared statement execution, thus less bugs.
@Chrriis , can you please update the PR with the updated branch and/or pass that through your tests?

I think the result is in a good for commit shape. As usual, more reviews are welcome.

Chrriis commented:
@vlsi I reviewed your changes, and I am very happy too about the removal of `statementType/statementName`. This is exactly what I was doing before I got side tracked by issue #584 :)
I also like all the renaming and cleanup you did. My main tests pass, but unfortunately I cannot do a complete validation because of the aforementioned issue.

Chrriis commented:
Following the workaround in issue #584 I ran again that test that got stuck.

It is a copy involving conversion of a 3.5 GB SQL Server database. It involves 281 tables, a handful of them having more than a million rows. The table on which it got stuck previously is the one with the highest number of rows (around 5 millions). For that test to pass, I had to lower MAX_BUFFERED_RECV_BYTES to 5000 (!).
The resulting column types I use are: NUMERIC (10), NUMERIC (19), SMALLINT, INTEGER, REAL, DOUBLE PRECISION, VARCHAR (7 precision variants), TEXT, BYTEA (using "stmt.setBinaryStream(xx)"), DATE, TIMESTAMP.
Indexes and foreign keys are added after actual copy.
The original copy took 12min 39s, with batch rewrite it takes 5min 19s.

I analyzed the results between copies made with and without rewrite. They both contain the same number of rows in resulting tables. I manually checked a few tables and they are identical. I launched the program that is using those databases, and everything it presented and processed was identical.

So, apart from the fact that it freezes (which is a showstopper), the actual rewriting feature works great!

vlsi commented:
>So, apart from the fact that it freezes (which is a showstopper), the actual rewriting feature works great!

Yeah, it looks like as rewriter manages to squeeze more data into a "single roundtrip", it makes the freeze   more likely to happen.

vlsi commented:
>In the code, I wrote:

Let me put that in another way: can you add a test that would fail unless `BatchedQuery. setStatementTypes` is overridden?

The test you mention does not fail at Travis. Neither does it fail at my OS X machine.

Chrriis commented:
I am not sure I follow. On my system, if I comment `BatchedQuery.setStatementTypes`, then the test `PreparedStatementPreparationTest` fails for `insertRewrite=true`. Of course the test did not fail in Travis, because there is the fix as part of my commit (the overridden `BatchedQuery.setStatementTypes`).

Am I missing something?

Chrriis commented:
Or maybe you mean a unit test of that particular method, rather than a more functional test like preparation? That I can do of course.

vlsi commented:
Look here: https://github.com/pgjdbc/pgjdbc/pull/585
It runs exactly the same test at current master and it does not hang.

Chrriis commented:
This test does not show the hang, it shows that only one preparation is done (S_1 rather than S_40). I just mentioned that the fact of not preparing statements again and again solved the freeze for my case.

vlsi commented:
Ah, I see, you have `getStatementName` test. Need to look closer. However, I still do not get what makes `BatchedQuery` so special so your magic fix applies only to BatchedQuery, not to the more generic SimpleQuery.

Chrriis commented:
It only applies to BatchedQuery because only that query has multiple times the same block in terms of types (cf `batchSize`), so I can try to find the common types that could apply to all blocks and repeat that.

vlsi commented:
> only that query has multiple times the same block in terms of types (cf batchSize),

Technically speaking, nothing prevents users from binding absolutely different values to the same bind position over and over again.
One should not expect that all blocks will end up in exactly the same type pattern.

vlsi commented:
On the other hand, datatype  of the same column does not change from one values block to another one. However, datatype reuse from "not-yet-described" statement is a slippery slope.

What if we introduce an artificial roundtrip right after the first prepare for "describe if unknown datatypes present"?

That is:
1) parse, describe, bind, exec;  SYNC
2) analyze describe result => update unknown bind types
3) continue with the rest of the batch

Chrriis commented:
I see something more troubling.

Assuming a table with 2 DATE columns,
If I do:
pstmt.setObject(1, null);
pstmt.setObject(2, ts);
pstmt.addBatch();
pstmt.executeBatch();

... I get types [1082, 1082].

But if I do:
pstmt.setNull(1, java.sql.Types.DATE);
pstmt.setObject(2, ts);
pstmt.addBatch();
pstmt.executeBatch();

... I get types [1082, 0].

It is the presence of this 0 that produces incompatible statement types (when I get [1082, 0], followed by [0, 1082]) and which I fixed with the override in BatchedQuery.
Is there a bigger issue here?

Chrriis commented:
Hm, I cannot put my hands on the exact conditions that produce this prepared statement rewrite. I need more eyes on this I guess. Just comment `BatchedQuery.setStatementTypes(int[])` and run the test `PreparedStatementPreparationTest`...

vlsi commented:
@Chrriis , there's a logic around here: https://github.com/pgjdbc/pgjdbc/blob/32d4e08f41e4c9fd5fbbeaefde3c364fe3f7f66d/pgjdbc/src/main/java/org/postgresql/jdbc/PgStatement.java#L972

It might be the logic of `org.postgresql.jdbc.PgPreparedStatement#isOneShotQuery` should account for batched queries (and enforce batch queries as `known-not-one-shot` ones)

Chrriis commented:
I understand the problem now.

If you change PreparedStatementPreparationTest ROW_INDEXES first value to 1, you also get the problem for a regular query (which means the test fails). While re-preparing a regular query may not be that much, the more parameters, the higher chances of having to re-prepare. Worse, the more parameters, the more data there is to send. This means higher pressure on the network interface resulting in my case in some freezes with batch rewriting.

So there are 2 points:
- My last commit is not necessary for the batch rewriting feature per se, so we should just consider the previous commit ("1f966e345edb9cee604a85a9d9cfbfdc1666e226") as a candidate for integration.
- We should solve this re-preparation problem and `PreparedStatementPreparationTest` could help in that (especially the tweak so that regular queries also fail).

What do you think? How to proceed?

vlsi commented:
>What do you think? How to proceed?

Here's the fix: https://github.com/vlsi/pgjdbc/commit/7197640f11a1eee3b1dee2d41f2e5af665c1e639#diff-e517b352b59b7d012c0a772c7e84d3eeR965

Can you try to test against your freeze scenario?

vlsi commented:
Here's the fix to "pre-describe the first statement in a batch"  https://github.com/vlsi/pgjdbc/commit/1d60fcf6a6d315e08c6797982cf9272fa6f8ca97#diff-e517b352b59b7d012c0a772c7e84d3eeR965

There is a room for improvement (e.g. being able to pre-describe in the middle of the batch), however we can implement that later.

Chrriis commented:
@vlsi I tested the heavy database copy test I mentioned earlier: fast, no freeze and result is as expected! So I reset my branch to your branch and forced push. I also modified the prepared statement test so that the regular batch is also tested (which passes now with your modifications).

Chrriis commented:
So you know, I tried a different test case and your pre-describe modifications also boost performance: I dropped from 28 seconds to 24 seconds. Great!

vlsi commented on pgjdbc/src/main/java/org/postgresql/core/v3/BatchedQuery.java:
```diff
@@ -0,0 +1,152 @@
+
+  @Override
+  public String toString() {
+    return getNativeSql();
```
`BatchedQuery.toString` should return not just SQL text, but the bind values as well.
It might be useful for debugging purposes (see how SimpleQuery returns SQL+binds)

Chrriis replied:
Hm, that is interesting. It is actually an aspect I hate about Postgres, for several reasons. My view is that the default should be the parameterized SQL, another method to get the parameters, and eventually another one with SQL + bind. I will come back on that another time, maybe with a pull request, who knows.
To come back to the current situation, should it show the latest bindings only?

Chrriis replied:
Oops, I got confused between statement and query String representation.
Still, looking at the code, I am not sure it shows SQL + binds. It seems to show "?" instead of "$1" though. Could we clarify exactly how it should be? Feel free to adjust the code if you want :)

vlsi replied:
> To come back to the current situation, should it show the latest bindings only?

Ah, we can end up with the same query object and different bind parameters in the same series of batch inserts.

Let's decide on that later then.

The idea with `toSting` is external profilers that instrument pgjdbc code to measure time and report sql/binds should have some sensible way of accessing query and binds without too much hard-coding for a specific pgjdbc version.
I do support an in-house profiler, and it is not a nice thing to reimplement "support of pgjdbc" as it gets upgraded. That is why I tried to rely on simple `toString`.

It might be time to introduce some proper API (that will cover sql, binds, batches, etc)

vlsi replied:
>Still, looking at the code, I am not sure it shows SQL + binds. It seems to show "?" instead of "$1" 

My bad. my local time is ~1am, and I should probably just have some sleep.

I use Query.toString for sql, and ParameterList for parameters. Current BatchQuery implementation should be just fine.

Chrriis replied:
> It might be time to introduce some proper API (that will cover sql, binds, batches, etc)

I like to hear that! I developped a low level (from our stack point of view) logger/debugger and it worked fine with other RDBMS because I could easily extract the SQL, the current set of parameters when adding the batch and a few other things. With Postgres we are blind. I added several layers of reflection but it is ugly and breaks at each upgrade. Worse: I cannot decode the parameters to see a float8 value for example. Last time I looked, the closest was logic coded within ResultSet...

vlsi replied:
> With Postgres we are blind

Feel free to raise an issue on that

Chrriis replied:
> Feel free to raise an issue on that

Done: #590 
Maybe the discussion from that issue will lead to a clear decision as to how statements/queries should behave. Then we can improve what we currently have.

vlsi commented on pgjdbc/src/main/java/org/postgresql/jdbc/BatchResultHandler.java:
```diff
@@ -149,4 +150,52 @@ public void handleCompletion() throws SQLException {
+    int offset = 0;
+    // In case of rewrite, we split into various queries
+    // It would be weird to have "no info" for some and 1 for another.
+    boolean isMultiValueBatchRewrite = queries[0].getBatchSize() > 1;
```
@Chrriis , are there any specific problems why you made `isMultiValueBatchRewrite` global instead of `per-query`?
Technically speaking, at this line, `hasRewrites==true`, so it is strange that you recalculate `isMultiValueBatchRewrite`.

I think of reverting to per-query `if (batchSize>1)`

Chrriis replied:
If I remember correctly, it was not a technical issue but rather nonsense from a user point of view.
1 batch = 1 row. Fine.
2 batches = 2 no info.
3 batches = 2 no info, 1 row.
...
6 batches = 6 no info.
7 batches = 6 no info, 1 row.

Good luck with explaining users that behavior :)

But using hasRewrites is probably better than isMultiValueBatchRewrite.

vlsi commented:
Just in case, here are some more measurements:

Medium batch sizes:
```
Benchmark                           (p1nrows)  (p2multi)  Mode  Cnt    Score    Error  Units
InsertBatch.insertBatchWithRewrite       1024          1  avgt   10    2,672 ±  0,185  ms/op
InsertBatch.insertBatchWithRewrite       1024          4  avgt   10    3,204 ±  0,911  ms/op
InsertBatch.insertBatchWithRewrite       1024          8  avgt   10    2,674 ±  0,236  ms/op
InsertBatch.insertBatchWithRewrite       1024         16  avgt   10    2,711 ±  0,281  ms/op
InsertBatch.insertBatchWithRewrite       1024        128  avgt   10    2,603 ±  0,185  ms/op
```

Big batch sizes: response time is comparable (yet we might consider batching more than 128 values), and "rely on rewriter" approach allocates more heap memory (that might be due ot SimpleParameterList objects being created for each `addBatch`)
```
Benchmark                                  (p1nrows)  (p2multi)  Mode  Cnt         Score      Error   Units
insertBatchWithRewrite                        102400          1  avgt    5       246,102 ±   64,468   ms/op
insertBatchWithRewrite:·gc.alloc.rate.norm    102400          1  avgt    5  44308466,080 ±  349,433    B/op
insertBatch                                   102400       1024  avgt    5       223,879 ±   47,590   ms/op
insertBatch:·gc.alloc.rate.norm               102400       1024  avgt    5  26063068,480 ±  279,578    B/op
insertBatch                                   102400          1  avgt    5       966,552 ±  161,170   ms/op
insertBatch:·gc.alloc.rate.norm               102400          1  avgt    5  58493524,800 ± 1625,625    B/op
```


Reference:
```
Benchmark                           (p1nrows)  (p2multi)  Mode  Cnt    Score    Error  Units
InsertBatch.insertBatch                    16        128  avgt   10    0,190 ±  0,020  ms/op
InsertBatch.insertBatch                   128        128  avgt   10    0,452 ±  0,049  ms/op
InsertBatch.insertBatch                  1024          1  avgt   10    9,530 ±  1,896  ms/op
InsertBatch.insertBatch                  1024          4  avgt   10    4,368 ±  0,359  ms/op
InsertBatch.insertBatch                  1024          8  avgt   10    3,283 ±  0,233  ms/op
InsertBatch.insertBatch                  1024         16  avgt   10    2,946 ±  0,477  ms/op
InsertBatch.insertBatch                  1024        128  avgt   10    2,519 ±  0,341  ms/op
InsertBatch.insertBatchWithRewrite         16        128  avgt   10    0,191 ±  0,026  ms/op
InsertBatch.insertBatchWithRewrite        128        128  avgt   10    0,442 ±  0,038  ms/op
InsertBatch.insertCopy                     16        128  avgt   10    0,206 ±  0,016  ms/op
InsertBatch.insertCopy                    128        128  avgt   10    0,330 ±  0,082  ms/op
InsertBatch.insertCopy                   1024          1  avgt   10  209,068 ± 15,828  ms/op
InsertBatch.insertCopy                   1024          4  avgt   10   53,216 ±  3,361  ms/op
InsertBatch.insertCopy                   1024          8  avgt   10   27,275 ±  2,722  ms/op
InsertBatch.insertCopy                   1024         16  avgt   10   14,829 ±  2,795  ms/op
InsertBatch.insertCopy                   1024        128  avgt   10    2,835 ±  0,647  ms/op
```

vlsi commented:
@Chrriis , many thanks for pushing this forward
